rfd/0098-securing-docker-images.md

```diff
@@ -0,0 +1,181 @@
+plugins. This would extend their use to Teleport proper.
+
+### 2. Ongoing scanning
+
+Using an _ongoing_ automated scanner means that we do not just check for 
+vulnerabilities at image creation time, instead we _continually_ scan for any vulnerabilities 
+that may be discovered during the image's lifetime.
+
+## Implementation Details
+
+## 1. Image construction
+
+### Teleport Image Requirements
+
+What is the minimal set of requirements to run Teleport on Linux in a container?
+
+   1. Teleport
+   2. `GLIBC` >= 2.17
+   3. `dumb-init` is required for correct signal and child processes handling
+      inside a container.
+   4. CA certificates
+
+Requirement (1) (i.e. Teleport itself) is provided by our CI process. 
+
+Requirements (2) and (4) are satisfied automatically by using the the google-
+provided base image [`gcr.io/distroless/cc-debian11`](https://github.com/GoogleContainerTools/distroless#what-images-are-available), which is configured for "mostly statically compiled" languages that require libc.
+
+Requirement (3) (`dumb-init`) can be sourced either from the upstream Debian repository, or
+downloaded directly from the project's GitHub release page. Sourcing `dumb-init` from the 
+Ubuntu or Debian package repositories implies some minimal provenance checking by the debian
+packaging tools, so we will prefer that to sourcing it from GitHub.
+
+### Base image verification
+
+The distroless base image will be pulled and verified prior to constructing the
+Teleport image, using the `cosign` tool as described [here](https://github.com/GoogleContainerTools/distroless#how-do-i-verify-distroless-images).
+
+This will allow us to specify a floating tag for the base image (and thus
+automatically include the latest version of the base image, with any security
+fixes, etc. included) while still validating the provenance of the base image
+itself.
+
+It is technically possible for the image to be poisoned post-validation (e.g.
+in a shared build environment, where a malicious peer could re-tag a malicious
+image as the base).
+
+While we _could_ verify that the validated base image appears in the parent 
+chain of the final image, this is still no protection against the malicious 
+image being based on the same parent. We could _also_ assert that the final 
+image's parent chain has an expecetd number of steps (inferred from the number
+of steps in the Docker file), but this is erorr-prone and would drastically 
+reduce the flexibility of the build system.
+
+For the above reasons, Teleport images for public consumption must not be 
+built in such a shared environment.
+
+### Building the image
+
+The image will be built from a multi-stage docker file, using build stages to download
+and unpack the required debian packages and copy them into place on the distroless 
+image. 
+
+An example Dockerfile, assuming the Teleport Debian package is supplied by the 
+CI system, might look like something like:
+
+```Docker
+FROM debian:11 as dumb-init
```

Comment:
How will we ensure we're staying on top of any security updates to `dumb-init` or any other one-off packages we install?